### DISCUSSION TITLE
REPL: Don't echo file on load but provide .listing command

### DISCUSSION THREAD

**Discussion by shonfeder** [informalsystems/quint#417] [Ideas]
Currently loading a file echos all the contents into the terminal. That seems a
bit noisy to me. Especially if you want to load several files in succession.
Instead, it'd be nice if loading a file ended with a simple success message, and
we provide a command like the Prolog top-level's `listing/1`, which will show
the current "database".

**Comment by konnov** [comment]
That sounds more like an issue to me.

**Reply by shonfeder** [reply]
I think I converted it from an issue, as I thought we only wanted issues for planned work, and we'd discuss the plans?

I'm not sure the line there. :smile: 

Sounds like you agree this would be useful?

**Reply by konnov** [reply]
Actually, it's a great idea! So `.load` would not print anything, but we could have something like `.print`?

Yeah, it reminds me a lot of `LIST` on my ZX Spectrum :)

**Reply by konnov** [reply]
Or maybe you have something like Python's inspect in mind: https://docs.python.org/3/library/inspect.html

**Reply by shonfeder** [reply]
> So .load would not print anything, but we could have something like .print?

Yeah, that's pretty much what I'm suggesting. Maybe you'd get:

```
>>> .load(`foo.qnt`)
File foo.qnt has been loaded.
```

But not the rest of the playback. If you wanted to see what was loaded, you'd issue another command.

Here's how it works in SWI-Prolog top-level, given a file in the cwd `foo.pl` containing

```prolog
:- module(foo, [bar/1]).

bar(a).
```

Then you can load it in the toplevel: 

```prolog
% listing the dir contents
?- ls.
% foo.pl

% equivalent to `.load`
?- consult(foo).
true.

% show just the contents added from the module `foot`
?- foo:listing.

bar(a).
true.
```

More generally

1. `?- listing.` Will print all the (non-stdlib) facts in the database.
2. `?- foo:listing.` Will print all the facts declared in module `foo`.
3. `?- listing(P/N).` Will print all the facts asserted for the predicate where `P` is an atom and `N` an arity.

This is just about printing stuff for a user to see the environment. I'm not sure if the is useful for us, but 1 and 2 probably are?

IIUC, Python's `inspect` is more a general purpose reflection: you can use it to examine and then operate on objects. That's not what I'm suggesting here.

**Comment by konnov** [comment]
Item 2 is interesting. I wonder about item 1 though. Would not it be easier for the user to call `.save foo.qnt` and read the file in their favorite editor?

**Reply by shonfeder** [reply]
> Would not it be easier for the user to call .save foo.qnt and read the file in their favorite editor?

Not in terms of the number of steps required for a user :p  

Given that we have `.save` maybe we don't need any printing functionality. I guess the main thrust of my suggestion here is that we don't echo back all the content of a file when we load it.

**Reply by konnov** [reply]
Yep, that's definitely a good improvement: https://github.com/informalsystems/tnt/issues/421

**Comment by konnov** [comment]
I don't know why this discussion is locked. It was probably meant to be closed.
